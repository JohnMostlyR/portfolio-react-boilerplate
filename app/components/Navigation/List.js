import PropTypes from 'prop-types';
import styled from 'styled-components';
import { lighten } from 'polished';

import { BASE_LINE_HEIGHT } from '../../styles/typography';
import { theme } from '../../styles/theme';
import { svgToURL } from '../../styles/tools';

const { backgroundColor, borderColor } = theme.site.navigation;
const lighterBorderColor =
  (borderColor && lighten(0.1, borderColor)) || 'inherit';
const arrowSVG = `
  <svg
    xmlns="http://www.w3.org/2000/svg"
    viewBox="0 0 848 640"
    preserveAspectRatio="xMaxYMax meet"
    >
    <path stroke="${borderColor}" fill="${borderColor}" d="M423.8,215.6L321.1,0v241.3" />
    <path fill="${lighterBorderColor}" d="M0,321.7l423.8-106.1v265.5" />
    <path fill="${borderColor}" d="M848,640L0,321.7L211.6,640H848z" />
  </svg>
`;

/* eslint-disable indent */
const List = styled.ul.attrs({
  hidden: ({ isExpanded }) => (isExpanded ? '"false"' : '"true"'),
})`
  display: flex;
  flex-flow: column nowrap;
  height: ${({ isExpanded }) => (isExpanded ? 'auto' : 0)};
  list-style: none;
  margin: 0;
  margin-top: ${BASE_LINE_HEIGHT}rem;
  position: relative;
  z-index: 2;
  border: 0.2rem solid ${borderColor};
  border-radius: 0.5rem;
  border-top-right-radius: 0;
  background-color: ${backgroundColor};
  padding: ${({ isExpanded }) =>
    isExpanded
      ? `${BASE_LINE_HEIGHT}rem ${BASE_LINE_HEIGHT}rem ${BASE_LINE_HEIGHT}rem ${BASE_LINE_HEIGHT *
          2}rem`
      : 0};
  transform: scale(${({ isExpanded }) => (isExpanded ? 1 : 0)});
  transform-origin: right top;
  transition: transform, padding, height;
  transition-duration: 0.1s;
  transition-timing-function: ease-in-out;

  &::before {
    position: absolute;
    content: '';
    height: ${BASE_LINE_HEIGHT}rem;
    left: 0;
    top: -0.15rem;
    transform: translateY(-100%);
    width: 100%;
    z-index: -1;
    background-image: ${() => `url(${svgToURL(arrowSVG)})`};
    background-position: right center;
    background-repeat: no-repeat;
    background-size: contain;
    color: inherit;
  }

  & li {
    display: ${({ isExpanded }) => (isExpanded ? 'inline-block' : 'none')};
  }

  & li:not(:last-child) {
    margin-bottom: ${BASE_LINE_HEIGHT}rem;
  }

  @media (min-width: ${({ bigScreenBreakpoint }) =>
      `${bigScreenBreakpoint}px`}) {
    height: auto;
    padding: ${`${BASE_LINE_HEIGHT}rem ${BASE_LINE_HEIGHT}rem ${BASE_LINE_HEIGHT}rem ${BASE_LINE_HEIGHT *
      2}rem`};
    transform: scale(1);
    width: auto;
    background-color: transparent;

    & li {
      display: inline-block;
    }
  }
`;

List.propTypes = {
  bigScreenBreakpoint: PropTypes.number.isRequired,
  isExpanded: PropTypes.bool,
};

export default List;
