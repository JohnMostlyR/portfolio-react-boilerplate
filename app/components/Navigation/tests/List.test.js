import React from 'react';
import { shallow } from 'enzyme';
import toJson from 'enzyme-to-json';
import 'jest-styled-components';

import List from '../List';

describe('<List />', () => {
  it('should render and match the snapshot', () => {
    const renderedComponent = shallow(<List bigScreenBreakpoint={1000} />);
    expect(toJson(renderedComponent)).toMatchSnapshot();
  });

  it('should adopt the "isExpanded" property', () => {
    const renderedComponent = shallow(
      <List bigScreenBreakpoint={1} isExpanded />,
    );
    expect(toJson(renderedComponent)).toMatchSnapshot();
  });
});
