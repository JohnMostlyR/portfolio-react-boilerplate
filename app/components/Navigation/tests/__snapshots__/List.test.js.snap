// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<List /> should adopt the "isExpanded" property 1`] = `
.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-flow: column nowrap;
  -ms-flex-flow: column nowrap;
  flex-flow: column nowrap;
  height: auto;
  list-style: none;
  margin: 0;
  margin-top: 1.5rem;
  position: relative;
  z-index: 2;
  border: 0.2rem solid #f63;
  border-radius: 0.5rem;
  border-top-right-radius: 0;
  background-color: rgba(0,0,0,0.8);
  padding: 1.5rem 1.5rem 1.5rem 3rem;
  -webkit-transform: scale(1);
  -ms-transform: scale(1);
  transform: scale(1);
  -webkit-transform-origin: right top;
  -ms-transform-origin: right top;
  transform-origin: right top;
  -webkit-transition: -webkit-transform,padding,height;
  -webkit-transition: transform,padding,height;
  transition: transform,padding,height;
  -webkit-transition-duration: 0.1s;
  transition-duration: 0.1s;
  -webkit-transition-timing-function: ease-in-out;
  transition-timing-function: ease-in-out;
}

.c0::before {
  position: absolute;
  content: '';
  height: 1.5rem;
  left: 0;
  top: -0.15rem;
  -webkit-transform: translateY(-100%);
  -ms-transform: translateY(-100%);
  transform: translateY(-100%);
  width: 100%;
  z-index: -1;
  background-image: url("data:image/svg+xml;charset=utf-8,  %3Csvg    xmlns='http://www.w3.org/2000/svg'    viewBox='0 0 848 640'    preserveAspectRatio='xMaxYMax meet'    %3E    %3Cpath stroke='%23f63' fill='%23f63' d='M423.8,215.6L321.1,0v241.3' /%3E    %3Cpath fill='%23ff8c66' d='M0,321.7l423.8-106.1v265.5' /%3E    %3Cpath fill='%23f63' d='M848,640L0,321.7L211.6,640H848z' /%3E  %3C/svg%3E");
  background-position: right center;
  background-repeat: no-repeat;
  background-size: contain;
  color: inherit;
}

.c0 li {
  display: inline-block;
}

.c0 li:not(:last-child) {
  margin-bottom: 1.5rem;
}

@media (min-width:1px) {
  .c0 {
    height: auto;
    padding: 1.5rem 1.5rem 1.5rem 3rem;
    -webkit-transform: scale(1);
    -ms-transform: scale(1);
    transform: scale(1);
    width: auto;
    background-color: transparent;
  }

  .c0 li {
    display: inline-block;
  }
}

<ul
  className="c0"
  hidden="\\"false\\""
/>
`;

exports[`<List /> should render and match the snapshot 1`] = `
.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-flow: column nowrap;
  -ms-flex-flow: column nowrap;
  flex-flow: column nowrap;
  height: 0;
  list-style: none;
  margin: 0;
  margin-top: 1.5rem;
  position: relative;
  z-index: 2;
  border: 0.2rem solid #f63;
  border-radius: 0.5rem;
  border-top-right-radius: 0;
  background-color: rgba(0,0,0,0.8);
  padding: 0;
  -webkit-transform: scale(0);
  -ms-transform: scale(0);
  transform: scale(0);
  -webkit-transform-origin: right top;
  -ms-transform-origin: right top;
  transform-origin: right top;
  -webkit-transition: -webkit-transform,padding,height;
  -webkit-transition: transform,padding,height;
  transition: transform,padding,height;
  -webkit-transition-duration: 0.1s;
  transition-duration: 0.1s;
  -webkit-transition-timing-function: ease-in-out;
  transition-timing-function: ease-in-out;
}

.c0::before {
  position: absolute;
  content: '';
  height: 1.5rem;
  left: 0;
  top: -0.15rem;
  -webkit-transform: translateY(-100%);
  -ms-transform: translateY(-100%);
  transform: translateY(-100%);
  width: 100%;
  z-index: -1;
  background-image: url("data:image/svg+xml;charset=utf-8,  %3Csvg    xmlns='http://www.w3.org/2000/svg'    viewBox='0 0 848 640'    preserveAspectRatio='xMaxYMax meet'    %3E    %3Cpath stroke='%23f63' fill='%23f63' d='M423.8,215.6L321.1,0v241.3' /%3E    %3Cpath fill='%23ff8c66' d='M0,321.7l423.8-106.1v265.5' /%3E    %3Cpath fill='%23f63' d='M848,640L0,321.7L211.6,640H848z' /%3E  %3C/svg%3E");
  background-position: right center;
  background-repeat: no-repeat;
  background-size: contain;
  color: inherit;
}

.c0 li {
  display: none;
}

.c0 li:not(:last-child) {
  margin-bottom: 1.5rem;
}

@media (min-width:1000px) {
  .c0 {
    height: auto;
    padding: 1.5rem 1.5rem 1.5rem 3rem;
    -webkit-transform: scale(1);
    -ms-transform: scale(1);
    transform: scale(1);
    width: auto;
    background-color: transparent;
  }

  .c0 li {
    display: inline-block;
  }
}

<ul
  className="c0"
  hidden="\\"true\\""
/>
`;
